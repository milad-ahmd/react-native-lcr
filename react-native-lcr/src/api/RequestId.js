/**
 *  The request message id used to call functions on the LCR
 */

const RequestId = {
  GetProductId: 0,
  GetFieldData: 32,
  SetFieldData: 33,
  GetDeliveryStatus: 40,
  IssueCommand: 36,
  AbortRequest: 126,
  CheckRequest: 125,
  PrintRequest: 34,
};

export default RequestId;
