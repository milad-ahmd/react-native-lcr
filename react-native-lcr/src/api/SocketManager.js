const tcp = require('react-native-tcp-socket');
const Buffer = require('buffer').Buffer;

const SOCKET_TIMEOUT = 2000;
const RECONNECT_DELAY = 3000;

const StreamStatus = {
  Disconnected: 'disconnected',
  Connected: 'connected',
  Connecting: 'connecting',
};

export default class SocketManager {
  constructor({onIdle, onLog, onStatusChange, onResponse}) {
    this.onIdle = onIdle;
    this.onLog = onLog;
    this.onStatusChange = onStatusChange;
    this.onResponse = onResponse;
    this.createSocket = this.createSocket.bind(this);
    this.connected = this.connected.bind(this);
    this.reconnect = this.reconnect.bind(this);
    this.connect = this.connect.bind(this);
    this.sendWriteBuffer = this.sendWriteBuffer.bind(this);
    this.socketValid = this.socketValid.bind(this);
  }

  write(message) {
    // if (this.writeBuffer) {
    //   this.onLog("Busy")
    // } else {
    this.writeBuffer = new Buffer(message, 'base64');
    this.sendWriteBuffer();
    // }
  }

  connect(environment) {
    this.environment = environment;
    this.connectedOnce = false;
    this.onLog('Connecting to: ' + environment.host + ':' + environment.port);
    this.createSocket();
  }

  createSocket() {
    try {
      if (this.socketValid()) {
        this.onLog('Socket already exists, reusing');
        this.onStatusChange(StreamStatus.Connected);
        return;
      }
      const {host, port} = this.environment;
      this.onStatusChange(StreamStatus.Connecting);

      let expired = false;
      const timer = setTimeout(() => {
        expired = true;
        this.onLog('Connection timeout');
        // this.socket.destroy();
        this.reconnect();
      }, SOCKET_TIMEOUT);

      this.socket = tcp.createConnection({host, port});
      this.socket.on('connect', () => {
        if (!expired) {
          this.discardSocket = false;
          clearTimeout(timer);
          this.onLog('Connected');
          this.connected();
        }
      });
      this.socket.on('error', error => {
        this.onLog('socket error event: ', error);
      });
    } catch (error) {
      this.onStatusChange(StreamStatus.Disconnected);
      this.onLog('Error creating socket: ' + error);
    }
  }

  connected() {
    if (!this.socketValid()) {
      this.onLog(
        'Error: connect() trying to add listeners to an unprepared socket',
      );
      return;
    }
    this.onStatusChange(StreamStatus.Connected);
    this.socket.on('close', () => this.reconnect());
    if (this.writeBuffer) {
      this.sendWriteBuffer();
    }
  }

  reconnect() {
    this.onLog(`Refreshing connection in ${RECONNECT_DELAY / 1000} seconds`);
    this.discardSocket = true;
    this.onStatusChange(StreamStatus.Disconnected);
    setTimeout(() => this.createSocket(), RECONNECT_DELAY); //delay a few seconds before retrying
  }

  socketValid() {
    return (
      this.socket &&
      !this.discardSocket &&
      !this.socket.destroyed &&
      !this.socket.connecting
    );
  }

  sendWriteBuffer() {
    if (!this.socketValid()) {
      return;
    }
    let buffer = new Buffer([]);
    let timer = null;
    const incDataTimeout = 300;

    this.socket.on('data', data => {
      clearTimeout(timer);
      buffer = Buffer.concat([buffer, data]);

      timer = setTimeout(() => {
        this.onResponse(buffer.toJSON().data);
        this.socket.removeAllListeners('data');
      }, incDataTimeout);
    });

    this.socket.write(this.writeBuffer, () => {
      this.writeBuffer = null;

      timer = setTimeout(() => {
        this.onIdle(true);
        this.socket.removeAllListeners('data');
      }, SOCKET_TIMEOUT);
    });
  }
}
