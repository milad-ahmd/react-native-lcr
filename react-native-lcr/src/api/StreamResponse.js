export class DataType {
  constructor(length = null) {
    this.length = length;
  }
}

export class AsciiType extends DataType {
  /**
   * Processes a byte as an ascii character
   *
   * @param {[Int]} bytes - An array of bytes represented as decimal
   */
  read(bytes) {
    var content = '';
    for (var i = 0; i < bytes.length; i++) {
      content += String.fromCharCode(bytes[i]);
    }
    return content;
  }
}

export class FloatType extends DataType {
  /**
   * Processes a byte as float
   *
   * @param {[Int]} bytes - An array of bytes represented as decimal
   */
  read(bytes) {
    // Create a buffer
    const buff = new ArrayBuffer(4);
    // Create a data view of it
    const view = new DataView(buff);

    // set bytes
    bytes.forEach(function (b, i) {
      view.setUint8(i, b);
    });
    const result = view.getFloat32(0).toFixed(2);
    return parseFloat(result);
  }
}

export class DecimalType extends DataType {
  /**
   * Processes a byte as decimal
   *
   * @param {[Int]} bytes - An array of bytes represented as decimal
   */
  read(bytes) {
    let binary = bytes.map(val => {
      let b = val.toString(2);
      // creates full 8 bit byte: toString(2) doesn't always include 8 bits
      let pad = new Array(8 - b.length + 1).join('0');
      return pad + b;
    });
    return parseInt(binary.join(''), 2);
  }
}

export class BitfieldType extends DataType {
  /**
   * Processes a byte as an array of bits
   */
  read(bytes) {
    let binary = bytes.map(byte => {
      let data = [];
      if (byte != null) {
        data = byte.toString(2).split('');
      }
      return new Array(8).fill(0).map(val => {
        return Number(data.pop() || 0) == 1;
      });
    });
    return [].concat(...binary.reverse()); //flatten
  }
}

export class StreamResponse {
  /**
   * Initializes a message.
   *
   * @constructor
   * @param {String} bytes - A decimal array containing the message contents in the following format:
   *                            ~~<to><from><status><len0><data0>...<datalen-1><crc0><crc1>
   * @param {(MessageDataType)} data - Message structure expected for data portion of the message
   */
  constructor(data = {}) {
    this.data = data;
  }

  read(bytes) {
    let prefix = new AsciiType().read([bytes[0], bytes[1]]);
    if (prefix != '~~') {
      return;
    }
    //bytes 0 and 1 represent "~~" prefix portion of message
    this.to = new DecimalType(1).read([bytes[2]]);
    this.from = new DecimalType(1).read([bytes[3]]);
    this.status = new BitfieldType(1).read([bytes[4]]);

    var message = bytes.slice(6, bytes.length - 2);
    const response = {};
    for (var key in this.data) {
      let type = this.data[key];
      if (type.length == null) {
        response[key] = type.read(message);
      } else {
        response[key] = type.read(message.splice(0, type.length));
      }
    }
    this.response = response;
    return response;
  }
}
