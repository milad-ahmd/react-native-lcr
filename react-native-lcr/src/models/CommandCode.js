/**
*   From LCP documentation:
*
*   Command Codes
*   Command   Description
*         0   Start/Resume a delivery.
*         1   Pause the current delivery.
*         2   End the current delivery and print the ticket if the settings allow.
*         3   Place the LCR in auxiliary state.
*         4   Place the LCR in shift state and print the ticket if settings allow.
*         5   Place the LCR in calibration state. This only applies if the LCR was previously in factory mode.
*         6   Print a ticket based on the current state of the LCR. This command does not end
*               any deliveries, nor will it print a ticket if a delivery is active.
*/

const CommandCode = {
  StartResume: 0,
  Pause: 1,
  Print: 2,
  ShiftPrint: 2,
}

export default CommandCode
