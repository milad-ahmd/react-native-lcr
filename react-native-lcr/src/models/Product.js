export default class Product {
  constructor(number, name) {
    this.number = number;
    this.name = name;
  }
}
