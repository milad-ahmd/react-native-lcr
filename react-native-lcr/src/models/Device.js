import Field from './Field';
import DeliveryCode from './DeliveryCode';
import {DecimalType, FloatType, AsciiType} from '../api/StreamResponse';

export default class Device {
  constructor(address) {
    this.address = address;
    this.deliveryCode = new DeliveryCode([]);
    this.products = {};

    // fields
    this.activeProductNumber = new Field(0, new DecimalType(1));
    this.activeProductDescriptor = new Field(11, new AsciiType());
    this.activePOSProductName = new Field(198, new AsciiType());
    this.activePOSProductNumber = new Field(197, new DecimalType(2));
    this.activePOSProductCalibration = new Field(201, new DecimalType(1));

    this.currentGrossQuantity = new Field(2, new DecimalType(4), 0);
    this.currentNetQuantity = new Field(3, new DecimalType(4), 0);
    this.currentFlowRate = new Field(4, new DecimalType(4), 0);
    this.calculatedTotalDue = new Field(230, new FloatType(4), 0);

    this.grossPreset = new Field(5, new DecimalType(4), 0);
    this.netPreset = new Field(6, new DecimalType(4), 0);
    this.pricePreset = new Field(192, new FloatType(4), 0);

    this.compensationType = new Field(57, new DecimalType(1), 0);
    this.volumeDecimalPlaces = new Field(39, new DecimalType(1), 1);
    this.pulsesPerUnit = new Field(47, new DecimalType());
    this.purchasedFeatures = new Field(125, new DecimalType());
  }

  deliveringPreset() {
    return this.activePresetField().value > 0 || this.pricePreset.value > 0;
  }

  activeQuantityField() {
    if (this.compensationType.value > 0) {
      return this.currentNetQuantity;
    }
    return this.currentGrossQuantity;
  }

  activePresetField() {
    if (this.compensationType.value > 0) {
      return this.netPreset;
    }
    return this.grossPreset;
  }

  posEnabled() {
    if (this.purchasedFeatures.value == null) {
      return false;
    }
    return this.purchasedFeatures.value == 1;
  }

  selectedProduct() {
    if (this.posEnabled()) {
      let num = this.activePOSProductNumber.value;
      return this.products[num];
    }
    let num = this.activeProductNumber.value;
    return this.products[num];
  }

  formattedQuantity() {
    return this.formattedDecimalType(this.activeQuantityField().value || 0);
  }

  formattedPresetValue() {
    return this.formattedDecimalType(this.activePresetField().value || 0);
  }

  formattedFlowRate() {
    return this.formattedDecimalType(this.currentFlowRate.value || 0);
  }

  formattedCalculatedTotalDue() {
    let total = this.calculatedTotalDue.value || 0;
    return '$' + total.toFixed(2);
  }

  formattedPricePresetValue() {
    let preset = this.pricePreset.value || 0;
    return '$' + preset.toFixed(2);
  }

  formattedDecimalType(amount) {
    let decimalPlaces = this.volumeDecimalPlaces.value || 1;
    let decimalAmount = amount / (10 * decimalPlaces);
    return decimalAmount.toFixed(decimalPlaces);
  }
}
