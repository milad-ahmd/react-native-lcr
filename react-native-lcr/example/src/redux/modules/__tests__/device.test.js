import {put, select} from 'redux-saga/effects';
import {
  actions as deviceActions,
  setFieldDataSaga,
  getAvailableProductsSaga,
  getDeliveryStatusSaga,
  initialState,
  reducer,
  getCurrentDeviceSelector,
} from '../device';
import snapshotTestModule from './snapshotTestModule';
import Device from '~/models/Device';
import Product from '~/models/Product';
import Field from '~/models/Field';
import CommandCode from '~/models/CommandCode';
import {DecimalType} from '~/api/StreamResponse';
import StreamQueueStub from '~/api/__tests__/StreamQueueStub';

const {
  clearDeliveryPresets,
  setDeliveryPreset,
  refreshDeliveryData,
  getFieldData,
  setFieldData,
  issueCommand,
  setActivePosProduct,
  setActiveProduct,
  getDeliveryStatus,
  getAvailableProducts,
  updateDevice,
  setSelectedDevice,
  setAvailableDevices,
  getAvailableDevices,
} = deviceActions;

const container = 'container name';
const device1 = new Device(1);
const device2 = new Device(2);
const devices = [device1, device2];
const field1 = new Field(1, new DecimalType(4), 0);
const field2 = new Field(2, new DecimalType(4), 0);
const fields = [field1, field2];
const product = new Product(2, 'product name');
const queue = new StreamQueueStub();
const isPricePreset = true;

snapshotTestModule({
  name: 'device',
  reducer,
  actions: {
    clearDeliveryPresets: clearDeliveryPresets(),
    setDeliveryPreset: setDeliveryPreset(container, isPricePreset, 30),
    refreshDeliveryData: refreshDeliveryData(),
    setActiveProduct: setActiveProduct(container, 1),
    getFieldData: getFieldData(container, fields),
    setFieldData: setFieldData(container, field1, 10, 20),
    issueCommand: issueCommand(container, CommandCode.StartResume),
    setActivePosProduct: setActivePosProduct(container, product),
    getDeliveryStatus: getDeliveryStatus(container),
    getAvailableProducts: getAvailableProducts(container),
    updateDevice: updateDevice(device1),
    setSelectedDevice: setSelectedDevice(container, device1),
    setAvailableDevices: setAvailableDevices(devices),
    getAvailableDevices: getAvailableDevices(),
  },
  resetState: () => {
    return {...initialState, availableDevices: devices};
  },
});

describe('saga', () => {
  const container = 'Screen';
  const device = new Device(1);

  describe('#setFieldDataSaga', () => {
    const field = device.currentGrossQuantity;
    field.value = 10;
    const action = setFieldData(container, field);
    const saga = setFieldDataSaga(action);

    it('fetches the selected device', () => {
      let result = saga.next();
      expect(result.value).toEqual(select(getCurrentDeviceSelector));
    });

    it('executes the request to set remote data', () => {
      let result = saga.next(device);
      expect(result.value.CALL.fn.name).toEqual('_setFieldData');
      expect(result.value.CALL.args).toEqual([device, field]);
    });
  });

  describe('#getAvailableProductsSaga', () => {
    let action = getAvailableProducts(container);
    let saga = getAvailableProductsSaga(action);

    it('fetches the selected device', () => {
      let result = saga.next();
      expect(result.value).toEqual(select(getCurrentDeviceSelector));
    });

    it('clears the existing products', () => {
      let result = saga.next(device);
      expect(result.value).toEqual(put(updateDevice(device)));
    });

    it('includes products with a pulses per unit value greater than 0', () => {
      for (var i = 0; i < 2; i++) {
        let result = saga.next();
        expect(result.value.CALL.fn.name).toEqual('_setFieldData');
        expect(result.value.CALL.args).toEqual([
          device,
          device.activeProductNumber,
        ]);

        result = saga.next();
        expect(result.value.CALL.fn.name).toEqual('_getFieldData');
        expect(result.value.CALL.args).toEqual([device, device.pulsesPerUnit]);

        result = saga.next();
        expect(result.value.CALL.fn.name).toEqual('_getFieldData');
        expect(result.value.CALL.args).toEqual([
          device,
          device.activeProductDescriptor,
        ]);
        result = saga.next();
        expect(result.value).toEqual(put(updateDevice(device)));
      }
    });

    it('omits products with a pulses per unit of 0', () => {
      device.pulsesPerUnit.value = 0;
      for (var i = 0; i < 2; i++) {
        let result = saga.next();
        expect(result.value.CALL.fn.name).toEqual('_setFieldData');
        expect(result.value.CALL.args).toEqual([
          device,
          device.activeProductNumber,
        ]);
        result = saga.next();
        expect(result.value.CALL.fn.name).toEqual('_getFieldData');
        expect(result.value.CALL.args).toEqual([device, device.pulsesPerUnit]);
      }
    });
  });

  describe('#getDeliveryStatusSaga', () => {
    const action = getDeliveryStatus(container, device);
    const saga = getDeliveryStatusSaga(action);

    it('fetches the selected device', () => {
      let result = saga.next();
      expect(result.value).toEqual(select(getCurrentDeviceSelector));
    });

    it('gets the delivery status', () => {
      let result = saga.next(device);
      expect(result.value.CALL.fn.name).toEqual('_getDeliveryStatus');
      expect(result.value.CALL.args).toEqual([device]);
    });

    it('puts the updateDevice action', () => {
      let result = saga.next();
      expect(result.value).toEqual(put(updateDevice(device)));
    });
  });
});
