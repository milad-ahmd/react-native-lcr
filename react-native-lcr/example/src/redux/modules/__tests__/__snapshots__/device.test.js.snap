// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action creators clearDeliveryPresets matches snapshot 1`] = `
Object {
  "type": "DEVICE/CLEAR_DELIVERY_PRESETS",
}
`;

exports[`action creators getAvailableDevices matches snapshot 1`] = `
Object {
  "payload": Object {
    "container": undefined,
  },
  "type": "DEVICE/GET_AVAILABLE_DEVICES",
}
`;

exports[`action creators getAvailableProducts matches snapshot 1`] = `
Object {
  "payload": Object {
    "container": "container name",
  },
  "type": "DEVICE/GET_AVAILABLE_PRODUCTS",
}
`;

exports[`action creators getDeliveryStatus matches snapshot 1`] = `
Object {
  "payload": Object {
    "container": "container name",
  },
  "type": "DEVICE/GET_DELIVERY_STATUS",
}
`;

exports[`action creators getFieldData matches snapshot 1`] = `
Object {
  "payload": Object {
    "container": "container name",
    "fields": Array [
      Field {
        "address": 1,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": undefined,
        "value": 0,
      },
      Field {
        "address": 2,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "currentGrossQuantity",
        "value": 0,
      },
    ],
  },
  "type": "DEVICE/GET_FIELD_DATA",
}
`;

exports[`action creators issueCommand matches snapshot 1`] = `
Object {
  "payload": Object {
    "command": 0,
    "container": "container name",
  },
  "type": "DEVICE/ISSUE_COMMAND",
}
`;

exports[`action creators refreshDeliveryData matches snapshot 1`] = `
Object {
  "type": "DEVICE/REFRESH_DELIVERY_DATA",
}
`;

exports[`action creators setActivePosProduct matches snapshot 1`] = `
Object {
  "payload": Object {
    "container": "container name",
    "product": Product {
      "name": "product name",
      "number": 2,
    },
  },
  "type": "DEVICE/SET_ACTIVE_POS_PRODUCT",
}
`;

exports[`action creators setActiveProduct matches snapshot 1`] = `
Object {
  "payload": Object {
    "container": "container name",
    "product": 1,
  },
  "type": "DEVICE/SET_ACTIVE_PRODUCT",
}
`;

exports[`action creators setAvailableDevices matches snapshot 1`] = `
Object {
  "payload": Object {
    "devices": Array [
      Device {
        "activePOSProductCalibration": Field {
          "address": 201,
          "dataType": DecimalType {
            "length": 1,
          },
          "name": "activePOSProductCalibration",
          "value": null,
        },
        "activePOSProductName": Field {
          "address": 198,
          "dataType": AsciiType {
            "length": null,
          },
          "name": "activePOSProductName",
          "value": null,
        },
        "activePOSProductNumber": Field {
          "address": 197,
          "dataType": DecimalType {
            "length": 2,
          },
          "name": "activePOSProductNumber",
          "value": null,
        },
        "activeProductDescriptor": Field {
          "address": 11,
          "dataType": AsciiType {
            "length": null,
          },
          "name": "activeProductDescriptor",
          "value": null,
        },
        "activeProductNumber": Field {
          "address": 0,
          "dataType": DecimalType {
            "length": 1,
          },
          "name": "activeProductNumber",
          "value": null,
        },
        "address": 1,
        "calculatedTotalDue": Field {
          "address": 230,
          "dataType": FloatType {
            "length": 4,
          },
          "name": "calculatedTotalDue",
          "value": 0,
        },
        "compensationType": Field {
          "address": 57,
          "dataType": DecimalType {
            "length": 1,
          },
          "name": "compensationType",
          "value": 0,
        },
        "currentFlowRate": Field {
          "address": 4,
          "dataType": DecimalType {
            "length": 4,
          },
          "name": "currentFlowRate",
          "value": 0,
        },
        "currentGrossQuantity": Field {
          "address": 2,
          "dataType": DecimalType {
            "length": 4,
          },
          "name": "currentGrossQuantity",
          "value": 0,
        },
        "currentNetQuantity": Field {
          "address": 3,
          "dataType": DecimalType {
            "length": 4,
          },
          "name": "currentNetQuantity",
          "value": 0,
        },
        "deliveryCode": DeliveryCode {
          "active": false,
          "deliveringGrossPreset": false,
          "deliveringNetPreset": false,
          "finished": undefined,
          "flowing": false,
          "shiftTicketPending": false,
          "ticketPending": false,
        },
        "grossPreset": Field {
          "address": 5,
          "dataType": DecimalType {
            "length": 4,
          },
          "name": "grossPreset",
          "value": 0,
        },
        "netPreset": Field {
          "address": 6,
          "dataType": DecimalType {
            "length": 4,
          },
          "name": "netPreset",
          "value": 0,
        },
        "pricePreset": Field {
          "address": 192,
          "dataType": FloatType {
            "length": 4,
          },
          "name": "pricePreset",
          "value": 0,
        },
        "products": Object {},
        "pulsesPerUnit": Field {
          "address": 47,
          "dataType": DecimalType {
            "length": null,
          },
          "name": "pulsesPerUnit",
          "value": null,
        },
        "purchasedFeatures": Field {
          "address": 125,
          "dataType": DecimalType {
            "length": null,
          },
          "name": "purchasedFeatures",
          "value": null,
        },
        "volumeDecimalPlaces": Field {
          "address": 39,
          "dataType": DecimalType {
            "length": 1,
          },
          "name": "volumeDecimalPlaces",
          "value": 1,
        },
      },
      Device {
        "activePOSProductCalibration": Field {
          "address": 201,
          "dataType": DecimalType {
            "length": 1,
          },
          "name": "activePOSProductCalibration",
          "value": null,
        },
        "activePOSProductName": Field {
          "address": 198,
          "dataType": AsciiType {
            "length": null,
          },
          "name": "activePOSProductName",
          "value": null,
        },
        "activePOSProductNumber": Field {
          "address": 197,
          "dataType": DecimalType {
            "length": 2,
          },
          "name": "activePOSProductNumber",
          "value": null,
        },
        "activeProductDescriptor": Field {
          "address": 11,
          "dataType": AsciiType {
            "length": null,
          },
          "name": "activeProductDescriptor",
          "value": null,
        },
        "activeProductNumber": Field {
          "address": 0,
          "dataType": DecimalType {
            "length": 1,
          },
          "name": "activeProductNumber",
          "value": null,
        },
        "address": 1,
        "calculatedTotalDue": Field {
          "address": 230,
          "dataType": FloatType {
            "length": 4,
          },
          "name": "calculatedTotalDue",
          "value": 0,
        },
        "compensationType": Field {
          "address": 57,
          "dataType": DecimalType {
            "length": 1,
          },
          "name": "compensationType",
          "value": 0,
        },
        "currentFlowRate": Field {
          "address": 4,
          "dataType": DecimalType {
            "length": 4,
          },
          "name": "currentFlowRate",
          "value": 0,
        },
        "currentGrossQuantity": Field {
          "address": 2,
          "dataType": DecimalType {
            "length": 4,
          },
          "name": "currentGrossQuantity",
          "value": 0,
        },
        "currentNetQuantity": Field {
          "address": 3,
          "dataType": DecimalType {
            "length": 4,
          },
          "name": "currentNetQuantity",
          "value": 0,
        },
        "deliveryCode": DeliveryCode {
          "active": false,
          "deliveringGrossPreset": false,
          "deliveringNetPreset": false,
          "finished": undefined,
          "flowing": false,
          "shiftTicketPending": false,
          "ticketPending": false,
        },
        "grossPreset": Field {
          "address": 5,
          "dataType": DecimalType {
            "length": 4,
          },
          "name": "grossPreset",
          "value": 0,
        },
        "netPreset": Field {
          "address": 6,
          "dataType": DecimalType {
            "length": 4,
          },
          "name": "netPreset",
          "value": 0,
        },
        "pricePreset": Field {
          "address": 192,
          "dataType": FloatType {
            "length": 4,
          },
          "name": "pricePreset",
          "value": 0,
        },
        "products": Object {},
        "pulsesPerUnit": Field {
          "address": 47,
          "dataType": DecimalType {
            "length": null,
          },
          "name": "pulsesPerUnit",
          "value": null,
        },
        "purchasedFeatures": Field {
          "address": 125,
          "dataType": DecimalType {
            "length": null,
          },
          "name": "purchasedFeatures",
          "value": null,
        },
        "volumeDecimalPlaces": Field {
          "address": 39,
          "dataType": DecimalType {
            "length": 1,
          },
          "name": "volumeDecimalPlaces",
          "value": 1,
        },
      },
    ],
  },
  "type": "DEVICE/SET_AVAILABLE_DEVICES",
}
`;

exports[`action creators setDeliveryPreset matches snapshot 1`] = `
Object {
  "payload": Object {
    "container": "container name",
    "isPricePreset": true,
    "value": 30,
  },
  "type": "DEVICE/SET_DELIVERY_PRESET",
}
`;

exports[`action creators setFieldData matches snapshot 1`] = `
Object {
  "payload": Object {
    "container": "container name",
    "data": 20,
    "field": Field {
      "address": 1,
      "dataType": DecimalType {
        "length": 4,
      },
      "name": undefined,
      "value": 0,
    },
    "localValue": 10,
  },
  "type": "DEVICE/SET_FIELD_DATA",
}
`;

exports[`action creators setSelectedDevice matches snapshot 1`] = `
Object {
  "payload": Object {
    "container": "container name",
    "device": Device {
      "activePOSProductCalibration": Field {
        "address": 201,
        "dataType": DecimalType {
          "length": 1,
        },
        "name": "activePOSProductCalibration",
        "value": null,
      },
      "activePOSProductName": Field {
        "address": 198,
        "dataType": AsciiType {
          "length": null,
        },
        "name": "activePOSProductName",
        "value": null,
      },
      "activePOSProductNumber": Field {
        "address": 197,
        "dataType": DecimalType {
          "length": 2,
        },
        "name": "activePOSProductNumber",
        "value": null,
      },
      "activeProductDescriptor": Field {
        "address": 11,
        "dataType": AsciiType {
          "length": null,
        },
        "name": "activeProductDescriptor",
        "value": null,
      },
      "activeProductNumber": Field {
        "address": 0,
        "dataType": DecimalType {
          "length": 1,
        },
        "name": "activeProductNumber",
        "value": null,
      },
      "address": 1,
      "calculatedTotalDue": Field {
        "address": 230,
        "dataType": FloatType {
          "length": 4,
        },
        "name": "calculatedTotalDue",
        "value": 0,
      },
      "compensationType": Field {
        "address": 57,
        "dataType": DecimalType {
          "length": 1,
        },
        "name": "compensationType",
        "value": 0,
      },
      "currentFlowRate": Field {
        "address": 4,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "currentFlowRate",
        "value": 0,
      },
      "currentGrossQuantity": Field {
        "address": 2,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "currentGrossQuantity",
        "value": 0,
      },
      "currentNetQuantity": Field {
        "address": 3,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "currentNetQuantity",
        "value": 0,
      },
      "deliveryCode": DeliveryCode {
        "active": false,
        "deliveringGrossPreset": false,
        "deliveringNetPreset": false,
        "finished": undefined,
        "flowing": false,
        "shiftTicketPending": false,
        "ticketPending": false,
      },
      "grossPreset": Field {
        "address": 5,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "grossPreset",
        "value": 0,
      },
      "netPreset": Field {
        "address": 6,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "netPreset",
        "value": 0,
      },
      "pricePreset": Field {
        "address": 192,
        "dataType": FloatType {
          "length": 4,
        },
        "name": "pricePreset",
        "value": 0,
      },
      "products": Object {},
      "pulsesPerUnit": Field {
        "address": 47,
        "dataType": DecimalType {
          "length": null,
        },
        "name": "pulsesPerUnit",
        "value": null,
      },
      "purchasedFeatures": Field {
        "address": 125,
        "dataType": DecimalType {
          "length": null,
        },
        "name": "purchasedFeatures",
        "value": null,
      },
      "volumeDecimalPlaces": Field {
        "address": 39,
        "dataType": DecimalType {
          "length": 1,
        },
        "name": "volumeDecimalPlaces",
        "value": 1,
      },
    },
  },
  "type": "DEVICE/SET_SELECTED_DEVICE",
}
`;

exports[`action creators updateDevice matches snapshot 1`] = `
Object {
  "payload": Object {
    "device": Device {
      "activePOSProductCalibration": Field {
        "address": 201,
        "dataType": DecimalType {
          "length": 1,
        },
        "name": "activePOSProductCalibration",
        "value": null,
      },
      "activePOSProductName": Field {
        "address": 198,
        "dataType": AsciiType {
          "length": null,
        },
        "name": "activePOSProductName",
        "value": null,
      },
      "activePOSProductNumber": Field {
        "address": 197,
        "dataType": DecimalType {
          "length": 2,
        },
        "name": "activePOSProductNumber",
        "value": null,
      },
      "activeProductDescriptor": Field {
        "address": 11,
        "dataType": AsciiType {
          "length": null,
        },
        "name": "activeProductDescriptor",
        "value": null,
      },
      "activeProductNumber": Field {
        "address": 0,
        "dataType": DecimalType {
          "length": 1,
        },
        "name": "activeProductNumber",
        "value": null,
      },
      "address": 1,
      "calculatedTotalDue": Field {
        "address": 230,
        "dataType": FloatType {
          "length": 4,
        },
        "name": "calculatedTotalDue",
        "value": 0,
      },
      "compensationType": Field {
        "address": 57,
        "dataType": DecimalType {
          "length": 1,
        },
        "name": "compensationType",
        "value": 0,
      },
      "currentFlowRate": Field {
        "address": 4,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "currentFlowRate",
        "value": 0,
      },
      "currentGrossQuantity": Field {
        "address": 2,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "currentGrossQuantity",
        "value": 0,
      },
      "currentNetQuantity": Field {
        "address": 3,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "currentNetQuantity",
        "value": 0,
      },
      "deliveryCode": DeliveryCode {
        "active": false,
        "deliveringGrossPreset": false,
        "deliveringNetPreset": false,
        "finished": undefined,
        "flowing": false,
        "shiftTicketPending": false,
        "ticketPending": false,
      },
      "grossPreset": Field {
        "address": 5,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "grossPreset",
        "value": 0,
      },
      "netPreset": Field {
        "address": 6,
        "dataType": DecimalType {
          "length": 4,
        },
        "name": "netPreset",
        "value": 0,
      },
      "pricePreset": Field {
        "address": 192,
        "dataType": FloatType {
          "length": 4,
        },
        "name": "pricePreset",
        "value": 0,
      },
      "products": Object {},
      "pulsesPerUnit": Field {
        "address": 47,
        "dataType": DecimalType {
          "length": null,
        },
        "name": "pulsesPerUnit",
        "value": null,
      },
      "purchasedFeatures": Field {
        "address": 125,
        "dataType": DecimalType {
          "length": null,
        },
        "name": "purchasedFeatures",
        "value": null,
      },
      "volumeDecimalPlaces": Field {
        "address": 39,
        "dataType": DecimalType {
          "length": 1,
        },
        "name": "volumeDecimalPlaces",
        "value": 1,
      },
    },
  },
  "type": "DEVICE/UPDATE_DEVICE",
}
`;

exports[`device module reducer #clearDeliveryPresets matches snapshot 1`] = `undefined`;

exports[`device module reducer #getAvailableDevices matches snapshot 1`] = `undefined`;

exports[`device module reducer #getAvailableProducts matches snapshot 1`] = `undefined`;

exports[`device module reducer #getDeliveryStatus matches snapshot 1`] = `undefined`;

exports[`device module reducer #getFieldData matches snapshot 1`] = `undefined`;

exports[`device module reducer #issueCommand matches snapshot 1`] = `undefined`;

exports[`device module reducer #refreshDeliveryData matches snapshot 1`] = `undefined`;

exports[`device module reducer #setActivePosProduct matches snapshot 1`] = `undefined`;

exports[`device module reducer #setActiveProduct matches snapshot 1`] = `undefined`;

exports[`device module reducer #setAvailableDevices matches snapshot 1`] = `undefined`;

exports[`device module reducer #setDeliveryPreset matches snapshot 1`] = `undefined`;

exports[`device module reducer #setFieldData matches snapshot 1`] = `undefined`;

exports[`device module reducer #setSelectedDevice matches snapshot 1`] = `
Array [
  DiffEdit {
    "kind": "E",
    "lhs": null,
    "path": Array [
      "selectedDeviceAddress",
    ],
    "rhs": 1,
  },
]
`;

exports[`device module reducer #updateDevice matches snapshot 1`] = `undefined`;
